### ISSUE TITLE
Assertion failure when selecting text from PDF background

### ISSUE THREAD

**Issue by rolandlo** [xournalpp/xournalpp#5343]
### Operating System

Linux

### (Linux only) Distribution

Ubuntu

### (Linux only) Desktop Environment

GNOME

### (Linux Only) Display Server

Wayland

### Installation Method

PPA / built from source

### Xournal++ Version

1.2.2+dev

### libgtk Version

3.24.38

### Bug Description

When selecting text with the PDF text element selection tool, Xournal++ crashes with a assertion error:
Assertion failure

```term
(com.github.xournalpp.xournalpp:92889): xopp-CRITICAL **: 16:13:56.199: Assertion failed: rv0 == static_cast<Float>(rv1)
    in function floor_cast
    at line 99 of /home/roland/dev/xournalpp/src/util/include/util/safe_casts.h
```
The issue is that 
- `PdfElemSelection::getRegionBbox()` may return an empty range, which has
```c++
    double minX = std::numeric_limits<double>::max();
    double minY = std::numeric_limits<double>::max();
    double maxX = std::numeric_limits<double>::lowest();
    double maxY = std::numeric_limits<double>::lowest();
```
- that range is passed to `XojPageView::flagDirtyRegion` and `XojPageView::repaintArea`
- the latter multiplies these number with `zoom`, so they become `inf` when `zoom > 1` , and then applies `floor_cast<int>` to the product
- In `floor_cast`:
```c++
    auto rv0 = std::floor(f);
    auto rv1 = Integral(rv0);
    xoj_assert(rv0 == static_cast<Float>(rv1));
```
`rv0` is `inf`, but `rv1` is `-2147483648` and the assumption fails, because the latter is not `inf` when casted back to a `double`

In other words `floor_cast` fails when `inf` is passed to it (which kind of makes sense; at least more sense than the result `-2147483648`). 

@Febbe @bhennion How should we fix that? It's very possible that the same issue appears in different places of the code, too.

### Expected Behaviour

No crash.

### Steps to Reproduce

Open a PDF and select text with the PDF text selection tool.

### Additional Context

Somehow the PPA is still compiled in debug mode...

**Comment by Febbe** [comment]
That's a very good catch of the cast. In my opinion we should never reach that code path, when PdfElemSelection::getRegionBbox() returns an infinite/no element containing region. Actually it should return either an optional or a 0,0,0,0 region. 
When looking at the code, it should be rewritten to something like:
1. init empty range to 0,0,0,0
2. for a in views::take(1) : reinit range to rectangle
3. for a in views::skip(1) : calculate min/max
Return range

**Comment by Febbe** [comment]
That's a very good catch of the cast. In my opinion we should never reach that code path, when PdfElemSelection::getRegionBbox() returns an infinite/no element containing region. Actually it should return either an optional or a 0,0,0,0 region. 
When looking at the code, it should be rewritten to something like:
1. init empty range to 0,0,0,0
2. for a in views::take(1) : reinit range to a
3. for a in views::skip(1) : calculate min/max(a, previous)
Return range

**Comment by rolandlo** [comment]
I think there are things that need to be checked with empty ranges. 
- Shouldn't a translated empty range be empty? It could become invalid, when translated by a large enough amount.
- Shouldn't an empty range with some added padding still be an empty range? It could also become invalid.
- What's the height and width of an empty range? Looks like `-inf` in both cases. Is that wanted?
- What should `XojPageView::flagDirtyRegion(const Range& rg)` do with an empty range? Is that something that shouldn't happen (so we can assert it is non-empty)?

**Comment by Febbe** [comment]
> I disagree. An empty Range is completely self explanatory, as opposed to {0,0,0,0}. We should however understand how this empty box is passed up to the view for repainting. For instance, the lines

Mathematically, any range which has a (height * width) == 0 is empty. The question is here, does it ever make sense to handle an empty range different from an invalid range (`!is_real(height * width)`).
My intent to just use {0,0,0,0} is, that nearly any operation on zero with another real number is still valid. This might hide issues, but does not cause crashes or UB, when the invalid range is passed even when it should not.

If we can handle empty ranges the same as invalid ranges, we can define them as any range which is empty, or its area is indeterminable (inf, nan, 0)

Currently, any non normalized range is defined as empty.
We should change that to:
`is_valid == normalized && height * width !=0 && is_real(height * width)`

> Those operations should indeed be dealt with with more care. Either we assume they are never called on empty ranges (e.g. add an assert(isValid())), or we fix them by adding a test.

I would go for an assertion in all functions which modify the range. A caller has to make sure, that he does not call those for invalid ranges.

> So the getWidth(), getHeight() could actually assert(isValid()), or return a std::optional. 

I would do neither of them, but using an assertion is better than the optional here. Using a nan or inf is never useful, 0 might be useless. But it's theoretically valid to return nan, inf or 0. 

> I would say it is the caller's responsiblity to ensure the passed on range is valid (hence non-empty).

I second that.

My proposal would be: 
  - to force the creation of Ranges from at least a single Point 
    - -> will prevent, that a Range is created in an invalid state.
  - every function which modifies the Range assert, that the range is valid after the operation.
  - when a Range can't be constructed, because a vector does not contain any points the function should return `std::nullopt` or skip dependent code paths.
  - define empty ranges as height * width == 0
  - define invalid ranges as !is_real(height * width)
